File: clean-code/code-ninja/src/main/java/com/codeninja/solid/StudentService.java
package com.codeninja.solid;

import java.io.Console;

public class StudentService {
	public Boolean addIfValid(String emailAddress, int universityId){
		//SRP
		System.out.printf("Log: Start add student with email %s", emailAddress);

		if (emailAddress.isEmpty()) 	{
			return false;
		}

		StudentRepository studentRepository = 
				new StudentRepository();

		if (studentRepository.exists(emailAddress))	{
			return false;
		}

		UniversityRepository universityRepository = 
				new UniversityRepository();

		University university = 
				universityRepository.getByID(universityId);

		Student student = new Student(emailAddress, universityId);

		//SRP, OCP
		PackageType type =
				university.getPackageType();
		if ( type == PackageType.Standard)
		{
			student.setMonthlyEbookAllowance(10);
		}
		else if (type == PackageType.Premium)
		{
			student.setMonthlyEbookAllowance (10 * 2);
		}

		studentRepository.add(student);

		//SRP
		System.out.printf("Log: End add student with email %s",emailAddress);

		return true;//CQS

	}
	
	public Student getStudentsByUniversity() {
		throw new NotImplementedException();
	}
	
	public Student getStudentsByCurrentlyBorrowedEbooks() {
		throw new NotImplementedException();
	}
}

File: code-ninja-course/CSHARP/Code/SOLIDSolution/Students/StudentService.cs
namespace LegacyCodeToTestableCode.SOLIDSolution
{
    public class StudentService
    {
        public void addIfValid(
                string emailAddress,
                int universityId,
                StudentRepository studentRepository,
                UniversityRepository universityRepository)
        {
            Logger.log("Start add student with email", emailAddress);

            validateEmail(emailAddress);

            if (studentRepository.exists(emailAddress))
            {
                return;
            }

            University university = universityRepository.getByID(universityId);
            Student student = studentRepository.createStudent(emailAddress, universityId);

            PackageType type = university.getPackageType();

            BookAllowanceSetter allowanceSetter = new BookAllowanceSetter(student);
            allowanceSetter.SetAllowance(type);

            studentRepository.add(student);

            Logger.log("End add student with email", emailAddress);

        }

        public Student getStudentsByUniversity()
        {
            throw new NotImplemented();
        }

        public Student getStudentsByCurrentlyBorrowedEbooks()
        {
            throw new NotImplemented();
        }

        private void validateEmail(string emailAddress)
        {
            if (emailAddress.IsEmpty())
                throw new InvalidEmail("Empty");
        }
    }

}
